mozilla-services/consvc-shepherd
# [SHEP-27] Rate limit adjusted GET & POST using retry-after header info

## References

[SHEP-27](https://mozilla-hub.atlassian.net/browse/SHEP-27)

## Problem Statement

Boostr's API has a 100 reqs per min rate limit that's inadequate for our syncing needs. During full syncs, we experience rate limits (429 errors). Our initial solution to introduce rests between each call  didn't always work. Likely due to how rate limits are handled by the Boostr API. 

## Proposed Changes

The 429 response includes a Retry-After header with the number of seconds to wait for before retrying. The new POST and GET functions uses this to determine how long to wait before retrying the sync. An extra second is added to the Retry-After time as an extra precaution. If no value is readable from the header, it will retry again in 60 seconds, which is enough time for the rate limit counter to fully reset.

## Verification Steps

Clear your DB tables and do a full sync (Approx run time should be 3.5 to 4 mins). If successful, there will be no 429 errors in the logs. 

[SHEP-27]: https://mozilla-hub.atlassian.net/browse/SHEP-27?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
